File: ports/go/src/gui/painting/bcspen.go
package painting

import (
	"sync"
	"github.com/robertpelloni/btk-go/ports/go/src/core/kernel"
)

type PenStyle int
const (
	NoPen PenStyle = iota
	SolidLine
	DashLine
	DotLine
	DashDotLine
	DashDotDotLine
	CustomDashLine
)

type PenCapStyle int
const (
	FlatCap PenCapStyle = iota
	SquareCap
	RoundCap
)

type PenJoinStyle int
const (
	MiterJoin PenJoinStyle = iota
	BevelJoin
	RoundJoin
	SvgMiterJoin
)

// BcsPen maps to QPen
type BcsPen struct {
	*kernel.BcsObject
	style     PenStyle
	color     string
	width     float64
	capStyle  PenCapStyle
	joinStyle PenJoinStyle
	mu        sync.RWMutex
}

func NewBcsPen() *BcsPen {
	return &BcsPen{
		BcsObject: kernel.NewBcsObject(nil),
		style:     SolidLine,
		color:     "#000000",
		width:     1.0,
		capStyle:  SquareCap,
		joinStyle: BevelJoin,
	}
}

func (p *BcsPen) SetStyle(s PenStyle) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.style = s
}

func (p *BcsPen) Style() PenStyle {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.style
}

func (p *BcsPen) SetColor(color string) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.color = color
}

func (p *BcsPen) Color() string {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.color
}

func (p *BcsPen) SetWidth(w float64) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.width = w
}

func (p *BcsPen) Width() float64 {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.width
}

func (p *BcsPen) SetCapStyle(c PenCapStyle) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.capStyle = c
}

func (p *BcsPen) CapStyle() PenCapStyle {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.capStyle
}

func (p *BcsPen) SetJoinStyle(j PenJoinStyle) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.joinStyle = j
}

func (p *BcsPen) JoinStyle() PenJoinStyle {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.joinStyle
}

File: ports/rust/src/gui/painting/bcs_pen.rs
use std::sync::{Arc, Mutex};
use crate::core::kernel::bcs_object::BcsObject;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PenStyle {
    NoPen = 0,
    SolidLine,
    DashLine,
    DotLine,
    DashDotLine,
    DashDotDotLine,
    CustomDashLine,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PenCapStyle {
    FlatCap = 0,
    SquareCap,
    RoundCap,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PenJoinStyle {
    MiterJoin = 0,
    BevelJoin,
    RoundJoin,
    SvgMiterJoin,
}

pub struct BcsPen {
    pub base: Arc<BcsObject>,
    style: Mutex<PenStyle>,
    color: Mutex<String>,
    width: Mutex<f64>,
    cap_style: Mutex<PenCapStyle>,
    join_style: Mutex<PenJoinStyle>,
}

impl BcsPen {
    pub fn new() -> Arc<Self> {
        Arc::new(Self {
            base: BcsObject::new(),
            style: Mutex::new(PenStyle::SolidLine),
            color: Mutex::new(String::from("#000000")),
            width: Mutex::new(1.0),
            cap_style: Mutex::new(PenCapStyle::SquareCap),
            join_style: Mutex::new(PenJoinStyle::BevelJoin),
        })
    }

    pub fn set_style(&self, style: PenStyle) {
        *self.style.lock().unwrap() = style;
    }

    pub fn style(&self) -> PenStyle {
        *self.style.lock().unwrap()
    }

    pub fn set_color(&self, color: &str) {
        *self.color.lock().unwrap() = color.to_string();
    }

    pub fn color(&self) -> String {
        self.color.lock().unwrap().clone()
    }

    pub fn set_width(&self, width: f64) {
        *self.width.lock().unwrap() = width;
    }

    pub fn width(&self) -> f64 {
        *self.width.lock().unwrap()
    }

    pub fn set_cap_style(&self, cap: PenCapStyle) {
        *self.cap_style.lock().unwrap() = cap;
    }

    pub fn cap_style(&self) -> PenCapStyle {
        *self.cap_style.lock().unwrap()
    }

    pub fn set_join_style(&self, join: PenJoinStyle) {
        *self.join_style.lock().unwrap() = join;
    }

    pub fn join_style(&self) -> PenJoinStyle {
        *self.join_style.lock().unwrap()
    }
}
